cli/command/telemetry_docker.go

```diff
@@ -0,0 +1,121 @@
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+const otelConfigFieldName = "otel"
+
+// dockerExporterOtlpEndpoint retrieves the OTLP endpoint used for the docker reporter
+// from the current context.
+func dockerExporterOtlpEndpoint(cli Cli) (endpoint string, secure bool, err error) {
+	meta, err := cli.ContextStore().GetMetadata(cli.CurrentContext())
+	if err != nil {
+		return "", false, err
+	}
+
+	var otelCfg any
+	switch m := meta.Metadata.(type) {
+	case DockerContext:
+		otelCfg = m.AdditionalFields[otelConfigFieldName]
+	case map[string]any:
+		otelCfg = m[otelConfigFieldName]
+	}
+
+	if otelCfg == nil {
+		return "", false, nil
+	}
+
+	otelMap, ok := otelCfg.(map[string]any)
+	if !ok {
+		return "", false, errors.Errorf(
+			"unexpected type for field %q: %T (expected: %T)",
+			otelConfigFieldName,
+			otelCfg,
+			otelMap,
+		)
+	}
+
+	// keys from https://opentelemetry.io/docs/concepts/sdk-configuration/otlp-exporter-configuration/
+	endpoint, ok = otelMap["OTEL_EXPORTER_OTLP_ENDPOINT"].(string)
+	if !ok {
+		return "", false, nil
+	}
+
+	// Parse the endpoint. The docker config expects the endpoint to be
+	// in the form of a URL to match the environment variable, but this
+	// option doesn't correspond directly to WithEndpoint.
+	//
+	// We pretend we're the same as the environment reader.
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", false, errors.Errorf("docker otel endpoint is invalid: %s", err)
+	}
+
+	switch u.Scheme {
+	case "unix":
+		// Unix sockets are a bit weird. OTEL seems to imply they
+		// can be used as an environment variable and are handled properly,
+		// but they don't seem to be as the behavior of the environment variable
+		// is to strip the scheme from the endpoint, but the underlying implementation
+		// needs the scheme to use the correct resolver.
+		//
+		// We'll just handle this in a special way and add the unix:// back to the endpoint.
+		endpoint = fmt.Sprintf("unix://%s", path.Join(u.Host, u.Path))
+	case "https":
+		secure = true
+		fallthrough
+	case "http":
+		endpoint = path.Join(u.Host, u.Path)
+	}
+	return endpoint, secure, nil
+}
+
+func dockerSpanExporter(cli Cli) (sdktrace.SpanExporter, error) {
+	endpoint, secure, err := dockerExporterOtlpEndpoint(cli)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := []otlptracegrpc.Option{
+		otlptracegrpc.WithEndpoint(endpoint),
+	}
+	if !secure {
+		opts = append(opts, otlptracegrpc.WithInsecure())
+	}
+
+	exp, err := otlptracegrpc.New(context.Background(), opts...)
+	if err != nil {
+		return nil, err
+	}
+	return exp, nil
+}
+
+func dockerMetricExporter(cli Cli) (sdkmetric.Exporter, error) {
+	endpoint, secure, err := dockerExporterOtlpEndpoint(cli)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := []otlpmetricgrpc.Option{
+		otlpmetricgrpc.WithEndpoint(endpoint),
+	}
+	if !secure {
+		opts = append(opts, otlpmetricgrpc.WithInsecure())
+	}
+
+	exp, err := otlpmetricgrpc.New(context.Background(), opts...)
```

Comment:
Shouldn't we accept a context from the function parameter instead of create a new one here?

Looking into the otel code base the `ctx` accepted here gets used inside the `grpc.DialContext(ctx, cfg.Metrics.Endpoint, dialOpts...)` function. 
https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlpmetric/otlpmetricgrpc/client.go#L57

For general cleanup purposes (cancel operation, timeouts etc.), I would prefer if we do not create standalone contexts.

Reply:
I'm not generally the biggest fan of these contexts used for initialization, but I've added them to this function just so we can pass it through. If the user wants to use the background context, they can always pass it in.

Reply:
I'm not too sure I follow? Wouldn't we want to have the ability to cancel, timeout the underlying otel code when say a `SIGINT` or `SIGTERM` is initiated by the user of the CLI?

Reading the docs of the `grpc.DialContext` function used by otel here, it _could_ be blocking which makes using a context the optimal way to cancel the dial. 

>In the blocking case, ctx can be used to cancel or expire the pending connection. Once this function returns, the cancellation and expiration of ctx will be noop. Users should call ClientConn.Close to terminate all the pending operations after this function returns. 
https://pkg.go.dev/google.golang.org/grpc#DialContext

Reply:
I took a deeper look at the code and I think this is fine. I didn't realize that the `New` invocation would initiate a dial. I thought it was a lazy invocation and the context was stored for later use.